Завдання 1

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
#1. Завантажте вбудований датасет load_breast_cancer з бібліотеки sklearn.datasets.

data = load_breast_cancer()
X = data.data
y = data.target
df = pd.DataFrame(X, columns=data.feature_names)
df['target'] = y

In [ ]:
# 2. Проведіть попередній аналіз даних:
#Перегляньте перші кілька рядків.
df.head()


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [ ]:
#Перевірте назви стовпців та типи даних.
df.dtypes

mean radius                float64
mean texture               float64
mean perimeter             float64
mean area                  float64
mean smoothness            float64
mean compactness           float64
mean concavity             float64
mean concave points        float64
mean symmetry              float64
mean fractal dimension     float64
radius error               float64
texture error              float64
perimeter error            float64
area error                 float64
smoothness error           float64
compactness error          float64
concavity error            float64
concave points error       float64
symmetry error             float64
fractal dimension error    float64
worst radius               float64
worst texture              float64
worst perimeter            float64
worst area                 float64
worst smoothness           float64
worst compactness          float64
worst concavity            float64
worst concave points       float64
worst symmetry             float64
worst fractal dimension    float64
target                       int64
dtype: object

In [ ]:
#Перевірте наявність пропущених значень.
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64

In [ ]:
#Досліджуйте розмір даних.
df.shape

(569, 31)

In [ ]:
#2. Побудова і налаштування моделей:
#Розділіть дані на навчальний і тестовий набори.
X, y = data.data, data.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [ ]:
#Побудуйте три моделі класифікації (на власний вибір).
# 1. Логістична регресія
logreg = LogisticRegression(max_iter=1000)
logreg_params = {
    'C': [0.1, 1, 10],
    'penalty': ['l2']
}

logreg_grid = GridSearchCV(logreg, logreg_params, cv=5)
logreg_grid.fit(X_train, y_train)
logreg_best = logreg_grid.best_estimator_

# 2. Дерево рішень
tree = DecisionTreeClassifier(random_state=42)
tree_params = {
    'max_depth': [3, 5, 7],
    'min_samples_split': [2, 5, 10]
}

tree_grid = GridSearchCV(tree, tree_params, cv=5)
tree_grid.fit(X_train, y_train)
tree_best = tree_grid.best_estimator_

# 3. Метод опорних векторів (SVM)
svm = SVC()
svm_params = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale', 'auto']
}

svm_grid = GridSearchCV(svm, svm_params, cv=5)
svm_grid.fit(X_train, y_train)
svm_best = svm_grid.best_estimator_

# Оцінка моделей
models = {'Logistic Regression': logreg_best, 'Decision Tree': tree_best, 'SVM': svm_best}

for model_name, model in models.items():
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{model_name} Accuracy: {accuracy}")

Logistic Regression Accuracy: 0.9824561403508771
Decision Tree Accuracy: 0.9649122807017544
SVM Accuracy: 0.9824561403508771


In [ ]:
#Підберіть оптимальні параметри для кожної моделі
best_params = {
    'Model': ['Logistic Regression', 'Decision Tree', 'SVM'],
    'Best Parameters': [
        logreg_grid.best_params_,
        tree_grid.best_params_,
        svm_grid.best_params_
    ]
}
best_params_df = pd.DataFrame(best_params)
best_params_df

Model                                   Best Parameters
0  Logistic Regression                         {'C': 1, 'penalty': 'l2'}
1        Decision Tree          {'max_depth': 3, 'min_samples_split': 2}
2                  SVM  {'C': 0.1, 'gamma': 'scale', 'kernel': 'linear'}

In [ ]:
#3.Оцініть моделі. Виведіть точність (accuracy) – обов’язково, матрицю помилок (confusion matrix) – за вибором та звіт про класифікацію (classification report) для кожної моделі.
# Обчислення точності (accuracy)
for model_name, model in models.items():
    print(f"Model: {model_name}")
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Точність (accuracy): {accuracy:.2f}")
    print('-' * 60)


Model: Logistic Regression
Точність (accuracy): 0.98
------------------------------------------------------------
Model: Decision Tree
Точність (accuracy): 0.96
------------------------------------------------------------
Model: SVM
Точність (accuracy): 0.98
------------------------------------------------------------


In [ ]:
# Виведення матриці помилок (confusion matrix)
for model_name, model in models.items():
    print(f"Model: {model_name}")
    y_pred = model.predict(X_test)
    conf_matrix = confusion_matrix(y_test, y_pred)
    print("Матриця помилок (confusion matrix):")
    print(conf_matrix)
    print('-' * 60)

Model: Logistic Regression
Матриця помилок (confusion matrix):
[[ 62   1]
 [  2 106]]
------------------------------------------------------------
Model: Decision Tree
Матриця помилок (confusion matrix):
[[ 59   4]
 [  2 106]]
------------------------------------------------------------
Model: SVM
Матриця помилок (confusion matrix):
[[ 61   2]
 [  1 107]]
------------------------------------------------------------


In [ ]:
# Виведення звіту про класифікацію (classification report)
for model_name, model in models.items():
    print(f"Model: {model_name}")
    y_pred = model.predict(X_test)
    class_report = classification_report(y_test, y_pred)
    print("Звіт про класифікацію (classification report):")
    print(class_report)
    print('-' * 60)

Model: Logistic Regression
Звіт про класифікацію (classification report):
              precision    recall  f1-score   support

           0       0.97      0.98      0.98        63
           1       0.99      0.98      0.99       108

    accuracy                           0.98       171
   macro avg       0.98      0.98      0.98       171
weighted avg       0.98      0.98      0.98       171

------------------------------------------------------------
Model: Decision Tree
Звіт про класифікацію (classification report):
              precision    recall  f1-score   support

           0       0.97      0.94      0.95        63
           1       0.96      0.98      0.97       108

    accuracy                           0.96       171
   macro avg       0.97      0.96      0.96       171
weighted avg       0.96      0.96      0.96       171

------------------------------------------------------------
Model: SVM
Звіт про класифікацію (classification report):
              precision 

In [44]:
#4. Прогнозування і висновки
best_params = {'C': 1, 'penalty': 'l2'}
logreg = LogisticRegression(**best_params, max_iter=1000)
logreg.fit(X_train, y_train)
np.random.seed(42)
random_indices = np.random.choice(X_test.shape[0], 10, replace=False)
X_random = X_test[random_indices]
y_random_actual = y_test[random_indices]
y_random_pred = logreg.predict(X_random)
comparison_df = pd.DataFrame({
    'Sample Index': random_indices,
    'Real Class': y_random_actual,
    'Predicted Class': y_random_pred
})
comparison_df

Sample Index  Real Class  Predicted Class
0           101           1                1
1            55           1                1
2            56           1                1
3           139           0                0
4           157           1                1
5            78           1                1
6           135           0                0
7           104           1                1
8           109           1                1
9           108           1                1

Висновок
Провели попередній аналіз даних. Розділили дані на навчальний і тестовий набори. Побудували три моделі класифікації: логістична регресія, дерево рішень та метод опорних векторів (SVM). Підібрали оптимальні параметри для кожної моделі. Найкраща модель на основі метрик: логістична регресія та SVM ці моделі мають найвищу точність(0.98).
